r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -2002,24 +2003,76 @@ test_that("`as_datetime()`", {
         ddate2 = lubridate::as_datetime(date),
         dchar_date_no_tz = as_datetime(char_date),
         dchar_date_with_tz = as_datetime(char_date, tz = "Pacific/Marquesas"),
+        dchar_date_subsec_no_tz = as_datetime(char_date_subsec),
+        dchar_date_subsec_with_tz = as_datetime(char_date_subsec, tz = "Pacific/Marquesas"),
         dint_date = as_datetime(int_date, origin = "1970-01-02"),
         dintegerish_date = as_datetime(integerish_date, origin = "1970-01-02"),
-        dintegerish_date2 = as_datetime(integerish_date, origin = "1970-01-01")
+        dintegerish_date2 = as_datetime(integerish_date, origin = "1970-01-01"),
+        ddouble_date = as_datetime(double_date)
       ) %>%
       collect(),
     test_df
   )
 
-  # Arrow does not support conversion of double to date
-  # the below should error with an error message originating in the C++ code
-  expect_error(
+  expect_identical(
     test_df %>%
       arrow_table() %>%
       mutate(
-        ddouble_date = as_datetime(double_date)
+        x = cast(as_datetime(double_date, unit = "ns"), int64()),
+        y = cast(as_datetime(double_date, unit = "us"), int64()),
+        z = cast(as_datetime(double_date, unit = "ms"), int64()),
+        .keep = "none"
+      ) %>%
+      collect(),
+    tibble(
+      x = bit64::as.integer64(c(10100000000, 25200000000, NA)),
+      y = as.integer(c(10100000, 25200000, NA)),
+      z = as.integer(c(10100, 25200, NA))
+    )
+  )
+})
+
+test_that("as_datetime() works with other functions", {
+  test_df <- tibble(
+    char_date = c("2022-03-22", "2021-07-30 14:32:47", "1970-01-01 00:00:59.123456789", NA)
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      transmute(
+        ddchar_date = as_datetime(char_date),
+        ddchar_date_date32_1 = as.Date(ddchar_date),
+        ddchar_date_date32_2 = as_date(ddchar_date),
+        ddchar_date_floored = floor_date(ddchar_date, unit = "days")
       ) %>%
       collect(),
-    regexp = "Float value 10.1 was truncated converting to int64"
+    test_df
+  )
+
+  # Arrow does not support conversion of timestamp to int32
+  # ARROW-17428
+  expect_error(
+    test_df %>%
+      arrow_table() %>%
+      mutate(
+        dchar_date = as_datetime(char_date),
+        dchar_date_int = as.integer(dchar_date)
+      ) %>%
+      compute(),
+    "Unsupported cast from timestamp\\[ns, tz=UTC\\] to int32 using function cast_int32"
+  )
+
+  # Arrow does not support conversion of timestamp to double
+  # ARROW-17428
+  expect_error(
+    test_df %>%
+      arrow_table() %>%
+      mutate(
+        dchar_date = as_datetime(char_date),
+        dchar_date_num = as.numeric(dchar_date)
+      ) %>%
+      compute(),
+    "Unsupported cast from timestamp\\[ns, tz=UTC\\] to double using function cast_double"
   )
```

Comment:
```suggestion
      collect()
  )
```
As the error comes from the C++ instead of R, we don't want to test the specific message contents here.